### PR TITLE
refactor: test steps into fns and make each flow 1 test

### PR DESCRIPTION
**Pull request by BlairCurrey**
## Changes:
- adds Peer to Peer - Cross Currency flow test
- adds Open Payments with Finish Method flow test (as opposed to the continuation polling)
- squashes each flow into 1 test - see the "Original PR body" collapse section below for more details.
- moves each step in the tests into a `admin` or `openPayments` test action. Makes things a bit more repeatable and the test a bit higher leve - see the "Original PR body" collapse section for more details.

## Context:

This is to merge into [bc/2529/cleanup-integration-test](https://github.com/interledger/rafiki/tree/bc/2529/cleanup-integration-tests), not main. I opened a separate PR in hopes of making these changes more digestable.

I originally opened this PR to get feedback on how I should refactor the tests (leave as-is, squash flows to single test but in-line everything, or squash flows to single test and turn the different steps into functions). I preserved that in the collapse section below, and it's what Max's comment is replying to here https://github.com/interledger/rafiki/pull/2608#issuecomment-2018997919:

### Original PR Body
<details>

<summary>details</summary>

My main question is: how should we structure the tests? I see a couple options:
1. Keep them how they are in main. That is, 1 test per request in the flow and sharing data between them via global scope.
2. 1 test per flow, with all requests/assertions in-lined and variables scoped to that test.
3. 1 test per flow with requests/assertions abstracted into repeatable functions that can be used across flows.

So for <span>#</span>2 or <span>#</span>3, our test would output like this instead of detailing each request in the flow.:

```
  Integration tests
    Requests
      ✓ Can Get Non-Existing Wallet Address (337 ms)
    Flows
      ✓ Open Payments with Continuation via Polling (1 ms)
      ✓ Open Payments with Continuation via "interact_ref"
      ✓ Peer to Peer (1374 ms)
      ✓ Peer to Peer - Cross Currency (1196 ms)
 ```

I'm pursuing <span>#</span>3, wondering if <span>#</span>2 might be better, and want to shared why I don't think <span>#</span>1 is a good option.

These changes start to implement <span>#</span>3.  I converted the `Peer to Peer` flow to 1 test and made a function for what were previously individual tests (`createReceiver`, `createQuote`, `createOutgoingPayment`, `getOutgoingPayment`). Then I implemented the `Peer to Peer - Cross Currency` flow test to see how reusable these functions were - I found that not too much had to be changed. Just the beginning and end of the flow basically. The input for `createReceiver` needs to be passed in wherever it's called, and some of the final assertions of the outgoing payments were flow specific so I pulled some assertions out to the callers. I think this is actually fine, but im wondering if the more these are re-used on different variations of the flows, the more we will have to pull out the assertions and spies and handle them where they are called. In which case this would really be over-abstracted as it would basically just wrap the gql/open payments client requests.

I'm doing this because I think option <span>#</span>1 is a problem basically because the different steps in the flow (create incoming payment, create quote, etc) are not independent. This means:
  - Flows should fail at the first failed test, which is not possible if each step is a test. With a test per flow, the first failed assertion will end the test and move to the next one (another flow). Whereas currently the next test is another step in the flow. No way around this in jest. It's just misusing the concept of tests.
  - We really only need to know if the flow passes/succeeds. We will still be able to see where it fails when it fails which is all that matters.
  - We have to thrust lots of data into the global scope so that tests can share results with subsequent test. This makes it hard to see exactly what each test relies on, and what side effects each test may have. This generally makes the tests a lot harder to understand.

I chose <span>#</span>3 over <span>#</span>2 because there would be a lot of code duplication for doing variations on flows (for example: `Open Payments with Continuation via Polling` and `Open Payments with Continuation via "interact_ref"`). However if it proves that we can't absrtact much of the underling api calls, spying/assertions then they aren't really reusable anyways I guess.

</details>

### CODE REVIEW

**Comment by mkurapov** [comment]
@BlairCurrey I think your intuition is right, #3 seems to be the best. 1 integration test = 1 jest test makes sense as well, since it encapsulates a single flow, and each step depends on the success of the previous. 

Likewise, abstracting into functions just makes individual test case easier to read as well :+1:

File: test/integration/lib/test-actions/index.ts
Comment (mkurapov) [inline test/integration/lib/test-actions/index.ts:38]: maybe we can pull these out into identity provider actions (or something along those lines)

File: test/integration/integration.test.ts
Comment (mkurapov) [reply test/integration/integration.test.ts:285]: > want to assume a certain conversion rate/fee structure was used

Since we have access to the config file, would we be able to get the rates from there and confirm the amounts that way?

**Review by mkurapov — APPROVED** [review]
LGTM, nothing blocking, just some thoughts

File: test/integration/lib/integration-server.ts
Comment (BlairCurrey) [reply test/integration/lib/integration-server.ts:4]: yeah I was using it it validate some id was a uuid but is was unecessary. It was used to generate a uuid but I used `crypto.randomUUID` instead. I guess normally it would be good to use the same uuid package everywhere but it was just for something throwaway in the test so i figured 1 less dependency was preferable.

File: test/integration/lib/test-actions/index.ts
Comment (BlairCurrey) [reply test/integration/lib/test-actions/index.ts:38]: We could. Although I was thinking it would be good to let these just be miscellaneous ones so we dont have to categorize everything. The admin api and open payments api are obvious but beyond that IDK how helpful it is to add the extra scope.

On a similar note, I originally was thinking I could do `AdminTestActions`, `OpenPaymentTestActions`, etc. instead of everything scoped to 1 `TestActions`. But then wondered where to throw these consent interaction ones. I guess `IdentityProviderTestActions` could be the answer but I didn't necessarily want to think of a scope for every potential new one.

File: test/integration/integration.test.ts
Comment (BlairCurrey) [reply test/integration/integration.test.ts:285]: I think there are fees too, so we'd need to get that from the seed as well. And then duplicate the non-trivial backend logic and live with the fact that it could change, or extract that out and use it here. What do you think? Seem like not great options IMO. I suppose maybe if really want to verify those amounts I can just hardcode what they currently are and state the assumptions in a comment (fee, rates, starting amount)? 

I guess im not sure, but perhaps those details aren't really that critical in this test? We're testing (or should be) those calculations and whether or not they're returned in the backend already. I think the important part of this test is that we can get the payment and see that its complete.

Comment (njlie) [reply test/integration/integration.test.ts:285]: Maybe checking that the conversion went in the right direction would be good enough? Like if we converted between currencies with a 2:1 ratio that the amount went down as opposed to up?

**Review by njlie — APPROVED** [review]

File: test/integration/integration.test.ts
Comment (mkurapov) [inline test/integration/integration.test.ts:300]: will this always be true though? depending on the fees, you might run into a scenario where the exchange rate is 0.98 or something, but with an added fee, this will fail this assertion. I guess what this means is that:

> I suppose maybe if really want to verify those amounts I can just hardcode what they currently are and state the assumptions in a comment (fee, rates, starting amount).

hardcoding the assumptions will always have to be the case.

With that said, I think we should either directly verify the amount, or not do the check at all. We can also leave it for a future enhancement, if you want. Thoughts?

Comment (BlairCurrey) [reply test/integration/integration.test.ts:300]: I thought of that so I changed the fee to 0, but I guess we would need to add some assertions checking the fee being 0. Which leads to you point about still having to hardcode assumptions.

I suppose making assertions (not jest `expect`, but `assert`) for fees, exchange rates, asset scale, asset code, and then expecting certain values is probably the better way to go. I'l include it here - this PR is still waiting on approval of the parent and the parents is waiting on the node 20 changes.

Comment (BlairCurrey) [reply test/integration/integration.test.ts]: add [0dda0e4](https://github.com/interledger/rafiki/pull/2608/commits/0dda0e41afc85503c9fc8b5a9864ad91f71141f4)

feel free to give anymore feedback on it but im merging this branch into the parent.

File: test/integration/integration.test.ts
Comment (BlairCurrey) [inline test/integration/integration.test.ts:293]: did some checks to ensure these are the operative seeds and fees.

**Pull request merged**
